File: cli/src/commands/verify.ts
import { Command } from 'commander';
import chalk from 'chalk';
import crypto from 'crypto';
import { formatOutput } from '../lib/utils';
import VeilPassAPI from '../lib/api';

function base64UrlDecode(str: string): string {
  return Buffer.from(str, 'base64url').toString('utf-8');
}

function padB64(s: string): string {
  return s + '='.repeat((4 - s.length % 4) % 4);
}

export const verifyCommand = new Command('verify')
  .description('Verify tokens, signed links, and signed URLs')
  .argument('<type>', 'Type to verify (token, link, url)')
  .argument('<value>', 'Value to verify')
  .option('--api <url>', 'API base URL for remote verification')
  .action(async (type: string, value: string, options: { api?: string }) => {
    try {
      if (options.api) {
        const api = new VeilPassAPI({ baseURL: options.api });
        const result = await api.verifyItem({ type, value });
        if (result.valid) {
          console.log(chalk.green('✓ Verification passed'));
        } else {
          console.log(chalk.red('✗ Verification failed'));
        }
        formatOutput(JSON.stringify(result, null, 2));
        return;
      }

      const now = Math.floor(Date.now() / 1000);
      let result: Record<string, unknown>;

      switch (type) {
        case 'token': {
          const parts = value.split('.');
          if (parts.length !== 3) throw new Error('Invalid JWT token format');
          let payload: Record<string, unknown>;
          try {
            payload = JSON.parse(base64UrlDecode(parts[1]));
          } catch {
            throw new Error('Failed to decode token payload');
          }
          const exp = payload.exp as number | undefined;
          const expired = exp ? now >= exp : false;
          result = {
            valid: !expired,
            type: 'token',
            expired,
            issuer: payload.iss || null,
            subject: payload.sub || null,
            audience: payload.aud || null,
            issued_at: payload.iat ? new Date((payload.iat as number) * 1000).toISOString() : null,
            expires_at: exp ? new Date(exp * 1000).toISOString() : null,
            claims: Object.keys(payload).filter(k => !['iss', 'sub', 'aud', 'iat', 'exp', 'jti', 'nbf'].includes(k)),
          };
          break;
        }
        case 'link': {
          const parts = value.split('.');
          if (parts.length !== 2) throw new Error('Invalid signed-link format (expected payload.signature)');
          let payload: Record<string, unknown>;
          try {
            const decoded = base64UrlDecode(padB64(parts[0]));
            payload = JSON.parse(decoded);
          } catch {
            throw new Error('Failed to decode signed-link payload');
          }
          const expStr = payload.exp as string | undefined;
          const expired = expStr ? new Date(expStr).getTime() < Date.now() : false;
          result = {
            valid: !expired,
            type: 'signed-link',
            expired,
            resource: payload.resource || null,
            one_time: payload.one_time || false,
            max_uses: payload.max_uses || null,
            issuer: payload.issuer || 'veilpass',
            issued_at: payload.iat || null,
            signature_present: !!parts[1],
          };
          break;
        }
        case 'url': {
          const parsed = new URL(value);
          const expires = parsed.searchParams.get('expires');
          const signature = parsed.searchParams.get('signature');
          const keyId = parsed.searchParams.get('key_id');
          const expired = expires ? parseInt(expires, 10) < now : false;
          result = {
            valid: !expired && !!signature,
            type: 'signed-url',
            expired,
            key_id: keyId || null,
            signature_present: !!signature,
            expires_at: expires ? new Date(parseInt(expires, 10) * 1000).toISOString() : null,
          };
          break;
        }
        default:
          throw new Error(`Unknown type: ${type}. Use: token, link, url`);
      }

      if (result.valid) {
        console.log(chalk.green('✓ Verification passed'));
      } else {
        console.log(chalk.red('✗ Verification failed'));
      }
      formatOutput(JSON.stringify(result, null, 2));
    } catch (err) {
      console.error(chalk.red('Verification error:'), (err as Error).message);
      process.exit(1);
    }
  });

File: apps/api/app/routes/verify.py
"""Verification endpoint for VeilPass API.

Supports verification of tokens, signed-links, signed-urls,
and SD-JWT (Selective Disclosure JWT) presentations.
"""

import base64
import hashlib
import json
from datetime import datetime, timezone
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from structlog import get_logger

from app.crypto import decode_jwt, ed25519_verify, hmac_verify, is_expired, validate_timestamp
from app.deps import request_id_dependency
from app.errors import InvalidInputError
from app.models.schemas import (
    SDJWTVeirfyResponse,
    VerifyRequest,
    VerifyResponse,
)
from app.routes.revoke import is_revoked
from app.sdjwt import verify_sd_jwt
from app.utils import _pad_b64

logger = get_logger(__name__)

router = APIRouter(prefix="/api/v1/verify", tags=["Verification"])


# ── Token identifier helpers ─────────────────────────────────────────


def _token_fingerprint(token: str) -> str:
    """Derive a stable identifier from a JWT for revocation lookups."""
    return hashlib.sha256(token.encode()).hexdigest()[:16]


def _check_revocation(token: str) -> None:
    """Raise InvalidInputError if the token has been revoked."""
    if is_revoked(_token_fingerprint(token)):
        raise InvalidInputError("Token has been revoked")


# ── Internal verifiers ───────────────────────────────────────────────


def _verify_token(value: str) -> dict:
    """Verify a JWT token and return claims."""
    _check_revocation(value)
    try:
        # Try Ed25519/EdDSA JWT verification
        decoded = decode_jwt(value)
        return {
            "valid": True,
            "expired": False,
            "issuer": decoded.get("iss", ""),
            "signature_valid": True,
            "claims": decoded,
        }
    except Exception as jwt_err:
        logger.warning("jwt_verification_failed", error=str(jwt_err))
        return {
            "valid": False,
            "expired": "expired" in str(jwt_err).lower(),
            "issuer": "",
            "signature_valid": False,
            "claims": {},
        }


def _verify_signed_link(value: str) -> dict:
    """Verify a signed-link token (payload.signature format)."""
    parts = value.split(".")
    if len(parts) != 2:
        return {
            "valid": False,
            "expired": False,
            "issuer": "veilpass",
            "signature_valid": False,
            "claims": {},
        }

    payload_b64, sig_b64 = parts
    try:
        # Decode payload
        payload_bytes = base64.urlsafe_b64decode(_pad_b64(payload_b64))
        payload = json.loads(payload_bytes.decode("utf-8"))
    except Exception:
        return {
            "valid": False,
            "expired": False,
            "issuer": "veilpass",
            "signature_valid": False,
            "claims": {},
        }

    # Verify signature — try Ed25519 first, fall back to HMAC
    signature_valid = False
    try:
        sig_bytes = base64.urlsafe_b64decode(_pad_b64(sig_b64))
        signature_valid = ed25519_verify(payload_bytes, sig_bytes)
    except Exception:
        try:
            signature_valid = hmac_verify(payload_bytes.decode("utf-8"), sig_b64)
        except Exception:
            signature_valid = False

    # Check expiration
    expired = False
    exp_str = payload.get("exp", "")
    if exp_str:
        try:
            exp_dt = datetime.fromisoformat(exp_str)
            if exp_dt.tzinfo is None:
                exp_dt = exp_dt.replace(tzinfo=timezone.utc)
            expired = datetime.now(timezone.utc) > exp_dt
        except (ValueError, TypeError):
            pass

    return {
        "valid": signature_valid and not expired,
        "expired": expired,
        "issuer": payload.get("issuer", "veilpass"),
        "signature_valid": signature_valid,
        "claims": payload,
    }


def _verify_signed_url(value: str) -> dict:
    """Verify a signed URL by checking its query parameters."""
    from urllib.parse import urlparse, parse_qs

    parsed = urlparse(value)
    params = parse_qs(parsed.query)

    expires = params.get("expires", [None])[0]
    signature = params.get("signature", [None])[0]
    key_id = params.get("key_id", [None])[0]

    if not expires or not signature:
        return {
            "valid": False,
            "expired": False,
            "issuer": "veilpass",
            "signature_valid": False,
            "claims": {},
        }

    # Check expiration
    try:
        exp_ts = int(expires)
        now_ts = int(datetime.now(timezone.utc).timestamp())
        expired = now_ts > exp_ts
    except (ValueError, TypeError):
        expired = False

    # Reconstruct the expected signature payload
    sig_payload = json.dumps(
        {
            "url": value,
            "expires": expires,
            "key_id": key_id,
        },
        separators=(",", ":"),
    )

    signature_valid = False
    try:
        sig_bytes = base64.urlsafe_b64decode(_pad_b64(signature))
        signature_valid = ed25519_verify(sig_payload.encode("utf-8"), sig_bytes)
    except Exception:
        try:
            signature_valid = hmac_verify(sig_payload, signature)
        except Exception:
            signature_valid = False

    return {
        "valid": signature_valid and not expired,
        "expired": expired,
        "issuer": "veilpass",
        "signature_valid": signature_valid,
        "claims": {"key_id": key_id, "expires": expires},
    }


# ── Standard verification endpoints ──────────────────────────────────


@router.post("", response_model=VerifyResponse, summary="Verify a token, signed-link, or signed-url")
async def verify_post(
    body: VerifyRequest,
    request_id: str = Depends(request_id_dependency),
):
    """Verify the validity of a token, signed-link, or signed-url."""
    if body.type == "token":
        result = _verify_token(body.value)
    elif body.type == "signed-link":
        result = _verify_signed_link(body.value)
    elif body.type == "signed-url":
        result = _verify_signed_url(body.value)
    else:
        raise InvalidInputError(f"Unsupported verification type: {body.type}")

    return VerifyResponse(
        valid=result["valid"],
        expired=result["expired"],
        issuer=result["issuer"],
        signature_valid=result["signature_valid"],
        claims=result["claims"],
        request_id=request_id,
    )


@router.get("", response_model=VerifyResponse, summary="Verify via query parameters")
async def verify_get(
    type: str = Query(..., pattern=r"^(token|signed-link|signed-url)$"),
    value: str = Query(..., max_length=8192),
    request_id: str = Depends(request_id_dependency),
):
    """Verify a token, signed-link, or signed-url via GET query parameters."""
    return await verify_post(
        body=VerifyRequest(type=type, value=value),
        request_id=request_id,
    )


# ── SD-JWT Verification ──────────────────────────────────────────────


class SDJWTVeirfyRequest(BaseModel):
    """Request model for SD-JWT verification."""

    token: str
    disclosures: list[str]


@router.post(
    "/sd-jwt",
    response_model=SDJWTVeirfyResponse,
    summary="Verify an SD-JWT presentation",
)
async def verify_sd_jwt_endpoint(
    body: SDJWTVeirfyRequest,
    request_id: str = Depends(request_id_dependency),
):
    """Verify an SD-JWT and return the disclosed claims.

    Provide the SD-JWT token and the disclosures the holder is presenting.
    The server verifies:
      1. The JWT signature (Ed25519/EdDSA)
      2. Token expiration
      3. Revocation status
      4. Each disclosure digest matches the embedded digest in the JWT

    On success, returns the verified disclosed claims.
    """
    if not body.token.strip():
        raise InvalidInputError("Token cannot be empty")

    if not body.disclosures:
        raise InvalidInputError("Disclosures list cannot be empty")

    # Check revocation
    _check_revocation(body.token)

    try:
        disclosed_claims = verify_sd_jwt(
            token=body.token,
            disclosures=body.disclosures,
        )
        return SDJWTVeirfyResponse(
            valid=True,
            expired=False,
            disclosed_claims=disclosed_claims,
            request_id=request_id,
        )
    except Exception as exc:
        err_str = str(exc).lower()
        is_expired = "expired" in err_str
        return SDJWTVeirfyResponse(
            valid=False,
            expired=is_expired,
            disclosed_claims={},
            request_id=request_id,
        )
